src/state/CallViewModel.ts

```diff
@@ -165,9 +307,193 @@ export class CallViewModel extends ViewModel {
+                    new UserMedia(userMediaId, member, p, this.encrypted),
+                ];
+
+                if (p.isScreenShareEnabled) {
+                  const screenShareId = `${userMediaId}:screen-share`;
+                  yield [
+                    screenShareId,
+                    prevItems.get(screenShareId) ??
+                      new ScreenShare(screenShareId, member, p, this.encrypted),
+                  ];
+                }
+              }
+            }.bind(this)(),
+          );
+
+          for (const [id, t] of prevItems) if (!newItems.has(id)) t.destroy();
+
+          // If every item is a ghost, that probably means we're still connecting
+          // and shouldn't bother showing anything yet
+          return allGhosts ? new Map() : newItems;
+        },
+        new Map<string, MediaItem>(),
+      ),
+      map((ms) => [...ms.values()]),
+      finalizeValue((ts) => {
+        for (const t of ts) t.destroy();
+      }),
+    ),
+  );
+
+  private readonly userMedia = this.mediaItems.pipe(
+    map((ms) => ms.filter((m): m is UserMedia => m instanceof UserMedia)),
+  );
+
+  private readonly screenShares = this.mediaItems.pipe(
+    map((ms) => ms.filter((m): m is ScreenShare => m instanceof ScreenShare)),
+  );
+
+  private readonly spotlightSpeaker = this.userMedia.pipe(
+    switchMap((ms) =>
+      combineLatest(
+        ms.map((m) => m.vm.speaking.pipe(map((s) => [m, s] as const))),
+      ),
+    ),
+    scan<(readonly [UserMedia, boolean])[], UserMedia | null, null>(
+      (prev, ms) =>
+        // Decide who to spotlight:
+        // If the previous speaker is still speaking, stick with them rather
+        // than switching eagerly to someone else
+        ms.find(([m, s]) => m === prev && s)?.[0] ??
+        // Otherwise, select anyone who is speaking
+        ms.find(([, s]) => s)?.[0] ??
+        // Otherwise, stick with the person who was last speaking
+        prev ??
+        // Otherwise, spotlight the local user
+        ms.find(([m]) => m.vm.local)?.[0] ??
+        null,
+      null,
+    ),
+    distinctUntilChanged(),
+    throttleTime(800, undefined, { leading: true, trailing: true }),
+  );
+
+  private readonly grid = this.userMedia.pipe(
+    switchMap((ms) => {
+      const bins = ms.map((m) =>
+        combineLatest(
+          [m.speaker, m.presenter, m.vm.audioEnabled, m.vm.videoEnabled],
+          (speaker, presenter, audio, video) =>
+            [
+              m,
+              m.vm.local
+                ? GridBin.SelfStart
+                : presenter
+                  ? GridBin.Presenters
+                  : speaker
+                    ? GridBin.Speakers
+                    : video
+                      ? audio
+                        ? GridBin.VideoAndAudio
+                        : GridBin.Video
+                      : audio
+                        ? GridBin.Audio
+                        : GridBin.NoMedia,
+            ] as const,
```

Comment:
Just a proposal/my opinion.
Sadly this is a lot longer but I find this easier to read:
```suggestion
    (speaker, presenter, audio, video) => {
        let binType = GridBin.SelfStart;
        if (m.vm.local) {
            binType = GridBin.SelfStart
        }
        else if (presenter) {
            binType = GridBin.Presenters
        }
        else if (speaker) {
            binType = GridBin.Speakers
        }
        else if (video && audio) {
            binType = GridBin.VideoAndAudio
        }
        else if (video && !audio) {
            binType = GridBin.Video
        }
        else if (!video && audio) {
            binType = GridBin.Audio
        }
        else if (!video && !audio) {
            binType = GridBin.NoMedia
        }
        return [m, binType] as const,
    }
```
(The ternaries need the additional processing step that usually the value after the ? is returned except in the end.)